### ISSUE TITLE
Error 500 when importing from MISP taxii

### ISSUE THREAD

**Issue by m1357** [MISP/MISP#5795]
# This template is meant for bug reports, if you have a feature request user the other template.

*If you would like to report a bug, please fill the template bellow*

### Work environment

| Questions                 | Answers
|---------------------------|--------------------
| Type of issue             | Bug
| OS version (server)       | Centos 7
| OS version (client)       | Centos 7
| PHP version               | 7.2
| MISP version / git hash   | v2.4.124 (0eb4e729702f9263632b994847d7ba2a48a3b7b6)
| Browser                   | If applicable

### Expected behavior
run test of taxii-push should import data.

### Actual behavior
Error 500

### Steps to reproduce the behavior
start gunicorn opentaxii.http:app --bind localhost:6000 --config python:opentaxii.http (or built in dev

Run 
taxii-push --path http://localhost:6000/services/inbox -f /opt/app/MISP-Taxii-Server/tests/test.xml --dest my_collection --username admin --password admin

### Logs, screenshots, configuration dump, ...
taxii server log:
2020-04-17T00:49:06.184791Z [__main__] info: Posting STIX...
2020-04-17T00:49:06.195008Z [__main__] info: Building Event...
2020-04-17T00:49:06.195454Z [__main__] info: Using title STIX Import
2020-04-17T00:49:06.200789Z [__main__] info: STIX loaded succesfully.
2020-04-17T00:49:06.200892Z [__main__] info: Extracted ['198.51.100.2', '198.51.100.17', '203.0.113.19']
2020-04-17T00:49:06.200978Z [__main__] info: Checking for existence of 198.51.100.2
2020-04-17T00:50:35.678072Z [werkzeug] info: 127.0.0.1 - - [16/Apr/2020 17:50:35] "POST /services/inbox HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 2450, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 1867, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib64/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)   
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib64/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/local/lib/python3.6/site-packages/opentaxii/middleware.py", line 85, in wrapper
    return _process_with_service(service)
  File "/usr/local/lib/python3.6/site-packages/opentaxii/middleware.py", line 163, in _process_with_service
    response_message = service.process(request.headers, taxii_message)
  File "/usr/local/lib/python3.6/site-packages/opentaxii/taxii/services/abstract.py", line 83, in process
    response_message = handler.handle_message(self, message)
  File "/usr/local/lib/python3.6/site-packages/opentaxii/taxii/services/handlers/inbox_message_handlers.py", line 123, in handle_message
    return InboxMessage11Handler.handle_message(service, request)
  File "/usr/local/lib/python3.6/site-packages/opentaxii/taxii/services/handlers/inbox_message_handlers.py", line 63, in handle_message
    inbox_message_id=inbox_message.id if inbox_message else None)
  File "/usr/local/lib/python3.6/site-packages/opentaxii/persistence/manager.py", line 197, in create_content
    collection_ids=collection_ids, service_id=service_id)
  File "/usr/local/lib/python3.6/site-packages/blinker/base.py", line 267, in send
    for receiver in self.receivers_for(sender)]
  File "/usr/local/lib/python3.6/site-packages/blinker/base.py", line 267, in <listcomp>
    for receiver in self.receivers_for(sender)]
  File "/usr/local/lib/python3.6/site-packages/misp_taxii_hooks-0.2-py3.6.egg/misp_taxii_hooks/hooks.py", line 69, in post_stix
    if search["response"]["Attribute"] != []:
KeyError: 'response'

################################################################33
Client logs :

Traceback (most recent call last):
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 2450, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 1867, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib64/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib64/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib64/python3.6/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/local/lib/python3.6/site-packages/opentaxii/middleware.py", line 85, in wrapper
    return _process_with_service(service)
  File "/usr/local/lib/python3.6/site-packages/opentaxii/middleware.py", line 163, in _process_with_service
    response_message = service.process(request.headers, taxii_message)
  File "/usr/local/lib/python3.6/site-packages/opentaxii/taxii/services/abstract.py", line 83, in process
    response_message = handler.handle_message(self, message)
  File "/usr/local/lib/python3.6/site-packages/opentaxii/taxii/services/handlers/inbox_message_handlers.py", line 123, in handle_message
    return InboxMessage11Handler.handle_message(service, request)
  File "/usr/local/lib/python3.6/site-packages/opentaxii/taxii/services/handlers/inbox_message_handlers.py", line 63, in handle_message
    inbox_message_id=inbox_message.id if inbox_message else None)
  File "/usr/local/lib/python3.6/site-packages/opentaxii/persistence/manager.py", line 197, in create_content
    collection_ids=collection_ids, service_id=service_id)
  File "/usr/local/lib/python3.6/site-packages/blinker/base.py", line 267, in send
    for receiver in self.receivers_for(sender)]
  File "/usr/local/lib/python3.6/site-packages/blinker/base.py", line 267, in <listcomp>
    for receiver in self.receivers_for(sender)]
  File "/usr/local/lib/python3.6/site-packages/misp_taxii_hooks-0.2-py3.6.egg/misp_taxii_hooks/hooks.py", line 69, in post_stix
    if search["response"]["Attribute"] != []:
KeyError: 'response'

-->

2020-04-16 18:01:59,918 cabby.cli.commons ERROR: HTTP Error: status code 500
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/cabby/cli/commons.py", line 171, in run_client
    run_func(client, args.uri, args)
  File "/usr/lib/python3.6/site-packages/cabby/cli/push.py", line 51, in _runner
    client.push(content, binding, collection_names=args.collections, uri=path)
  File "/usr/lib/python3.6/site-packages/cabby/client11.py", line 332, in push
    service_type=const.SVC_INBOX)
  File "/usr/lib/python3.6/site-packages/cabby/abstract.py", line 190, in _execute_request
    timeout=self.timeout)
  File "/usr/lib/python3.6/site-packages/cabby/dispatcher.py", line 54, in send_taxii_request
    stream, headers = request_stream(session, url, request_body, timeout)
  File "/usr/lib/python3.6/site-packages/cabby/dispatcher.py", line 91, in request_stream
    raise_http_error(response.status_code, response.raw)
  File "/usr/lib/python3.6/site-packages/cabby/dispatcher.py", line 34, in raise_http_error
    raise HTTPError(status_code)
cabby.exceptions.HTTPError: HTTP Error: status code 500

Please let me know what else I can provide.

**Comment by m1357** [comment]
Gunicorn output:

gunicorn opentaxii.http:app --bind 0.0.0.0:6000 --config python:opentaxii.http
{"api": "opentaxii.persistence.sqldb.SQLDatabaseAPI", "event": "api.initialized", "logger": "opentaxii.utils", "level": "info", "timestamp": "2020-04-17T01:23:56.145040Z"}
{"api": "opentaxii.auth.sqldb.SQLDatabaseAPI", "event": "api.initialized", "logger": "opentaxii.utils", "level": "info", "timestamp": "2020-04-17T01:23:56.166044Z"}
Using config from /opt/app/MISP-Taxii-Server/config/config.yaml
{"hooks": "misp_taxii_hooks.hooks", "event": "signal_hooks.imported", "logger": "opentaxii.server", "level": "info", "timestamp": "2020-04-17T01:23:56.504590Z"}
{"event": "opentaxii.server_configured", "logger": "opentaxii.server", "level": "info", "timestamp": "2020-04-17T01:23:56.504829Z"}
[2020-04-16 18:23:56 -0700] [11107] [INFO] Starting gunicorn 20.0.4
[2020-04-16 18:23:56 -0700] [11107] [INFO] Listening at: http://0.0.0.0:6000 (11107)
[2020-04-16 18:23:56 -0700] [11107] [INFO] Using worker: sync
[2020-04-16 18:23:56 -0700] [11118] [INFO] Booting worker with pid: 11118
[2020-04-16 18:24:17 -0700] [11118] [INFO] Posting STIX...
[2020-04-16 18:24:17 -0700] [11118] [INFO] Building Event...
[2020-04-16 18:24:17 -0700] [11118] [INFO] Using title STIX Import
[2020-04-16 18:24:17 -0700] [11118] [INFO] STIX loaded succesfully.
[2020-04-16 18:24:17 -0700] [11118] [INFO] Extracted ['198.51.100.2', '198.51.100.17', '203.0.113.19']
[2020-04-16 18:24:17 -0700] [11118] [INFO] Checking for existence of 198.51.100.2
[2020-04-16 18:24:47 -0700] [11107] [CRITICAL] WORKER TIMEOUT (pid:11118)
[2020-04-16 18:24:47 -0700] [11118] [INFO] Worker exiting (pid: 11118)
[2020-04-16 18:24:47 -0700] [11169] [INFO] Booting worker with pid: 11169

**Comment by ruffy91** [comment]
I am also using https://github.com/MISP/MISP-Taxii-Server to push MISP events to OpenTaxii and have the same error

Questions | Answers
-- | --
Type of issue | Bug
OS version (server) | Centos 7
OS version (client) | Centos 7
PHP version | 7.2
MISP version / git hash | v2.4.124 (0eb4e72)
Browser | If applicable

Logs, screenshots, configuration dump, ...

Apr 20 09:13:52 RB00SRV032 gunicorn[30222]: {"event": "Checking for existence of 101.55.125.10", "logger": "main", "level": "info", "timestamp": "2020-04-20T09:13:52.470809Z"}
Apr 20 09:13:52 RB00SRV032 gunicorn[30222]: {"event": "Exception on /services/inbox [POST]", "exc_info": ["<class 'KeyError'>", "KeyError('response',)", "<traceback object at 0x7f7fb73fa208>"], "logger": "opentaxii.middleware", "level": "error", "timestamp": "2020-04-20T09:13:52.975626Z"}
Apr 20 09:13:52 RB00SRV032 gunicorn[30222]: {"event": "Internal error", "logger": "opentaxii.middleware", "level": "error", "timestamp": "2020-04-20T09:13:52.976318Z", "exception": "Traceback (most recent call last):
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/flask/app.py", line 2446, in wsgi_app
 response = self.full_dispatch_request()
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/flask/app.py", line 1951, in full_dispatch_request
 rv = self.handle_user_exception(e)
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/flask/app.py", line 1820, in handle_user_exception
 reraise(exc_type, exc_value, tb)
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
 raise value
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/flask/app.py", line 1949, in full_dispatch_request
 rv = self.dispatch_request()
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/flask/app.py", line 1935, in dispatch_request
 return self.view_functionsrule.endpoint
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/opentaxii/middleware.py", line 85, in wrapper
 return _process_with_service(service)
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/opentaxii/middleware.py", line 163, in _process_with_service
 response_message = service.process(request.headers, taxii_message)
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/opentaxii/taxii/services/abstract.py", line 83, in process
 response_message = handler.handle_message(self, message)
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/opentaxii/taxii/services/handlers/inbox_message_handlers.py", line 123, in handle_message
 return InboxMessage11Handler.handle_message(service, request)
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/opentaxii/taxii/services/handlers/inbox_message_handlers.py", line 63, in handle_message
 inbox_message_id=inbox_message.id if inbox_message else None)
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/opentaxii/persistence/manager.py", line 197, in create_content
 collection_ids=collection_ids, service_id=service_id)
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/blinker/base.py", line 267, in send
 for receiver in self.receivers_for(sender)]
 File "/opt/MISP-Taxii-Server/venv/lib/python3.6/site-packages/blinker/base.py", line 267, in 
 for receiver in self.receivers_for(sender)]
 File "/opt/MISP-Taxii-Server/misp_taxii_hooks/hooks.py", line 69, in post_stix
 if search["response"]["Attribute"] != []:
KeyError: 'response'"}

**Comment by m1357** [comment]
I have got this working. a few things I did but not exactly sure where the root issue was. There still seem to be some default ownership settings using www-data (debian/ubuntu) user:
chown -R apache:apache /var/www/MISP

faup seemed to be missing 
Installed faup from source and install python bindings:
sudo -H -u apache /var/www/MISP/venv/bin/python faup/src/lib/bindings/python/setup.py install
Rebuild misp modules:
sudo -H -u apache /var/www/MISP/venv/bin/pip3 install -r REQUIREMENTS
sudo -H -u apache /var/www/MISP/venv/bin/pip3 install -U .
systemd still starts with www-data users

misp systemd scripts start with www-data user so change that to apache in:
/etc/systemd/system/misp-modules.service and misp-workers.service

I'm not certain on the ownership of all the files but everything seems to be working when owned as apache.
